include/rmm/device_scalar.hpp

```diff
@@ -36,75 +38,152 @@ class device_scalar {
+   * or on another stream only if a dependency is enforced (e.g. using `cudaStreamWaitEvent()`).
+   * 
+   * @throws `rmm::bad_alloc` if allocating the device memory for `initial_value` fails.
+   * @throws `rmm::cuda_error` if copying `initial_value` to device memory fails.
+   *
+   * @param initial_value The initial value of the object in device memory.
+   * @param stream Optional, stream on which to perform allocation and copy.
+   * @param mr Optional, resource with which to allocate.
    */
   explicit device_scalar(
-      T const &initial_value, cudaStream_t stream = 0,
-      rmm::mr::device_memory_resource *mr = rmm::mr::get_default_resource())
-      : buff{sizeof(T), stream, mr} {
-    _memcpy(buff.data(), &initial_value, stream);
+      T const& initial_value,
+      cudaStream_t stream = 0,
+      rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource())
+      : buffer{sizeof(T), stream, mr} {
+    set_value(initial_value, stream);
   }
 
   /**
-   * @brief Copies the value from device to host, synchronizes, and returns the
-   * value.
+   * @brief Copies the value from device to host, synchronizes, and returns the value.
+   * 
+   * Synchronizes `stream` after copying the data from device to host.
+   * 
+   * @note If the stream specified to this function is different from the stream specified
+   * to the constructor, then an appropriate dependency must be inserted between the streams 
+   * (e.g. using `cudaStreamWaitEvent()` or `cudaStreamSynchronize()`) before calling this function,
+   * otherwise there may be a race condition.
    *
-   * @throws `rmm::cuda_error` If the copy fails
-   * @throws `rmm::cuda_error` If synchronizing `stream` fails
+   * @throws `rmm::cuda_error` If the copy fails.
+   * @throws `rmm::cuda_error` If synchronizing `stream` fails.
    *
-   * @return T The value of the scalar after synchronizing the stream
-   * @param stream CUDA stream on which to perform the copy
+   * @return T The value of the scalar.
+   * @param stream CUDA stream on which to perform the copy and synchronize.
    */
   T value(cudaStream_t stream = 0) const {
     T host_value{};
-    _memcpy(&host_value, buff.data(), stream);
+    _memcpy(&host_value, buffer.data(), stream);
+    RMM_CUDA_TRY(cudaStreamSynchronize(stream));
     return host_value;
   }
 
   /**
-   * @brief Copies the value from host to device and synchronizes the stream.
+   * @brief Sets the value of the `device_scalar` to the given `host_value`.
+   *
+   * @note If the stream specified to this function is different from the stream specified
+   * to the constructor, then appropriate dependencies must be inserted between the streams 
+   * (e.g. using `cudaStreamWaitEvent()` or `cudaStreamSynchronize()`) before and after calling
+   * this function, otherwise there may be a race condition.
    *
+   * Does not synchronize `stream`.
+   * 
    * @throws `rmm::cuda_error` if copying `host_value` to device memory fails
```

Comment:
```suggestion
   * @throws `rmm::cuda_error` if copying `host_value` to device memory fails.
```
Since you were adding `.` above. I still see some throws and param lines below that don't end with `.` either.